### PR TITLE
merged_tree: split function that merges trees without resolving conflicts

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/rewrite.rs [hunk 1]
```diff
@@ -58,9 +64,10 @@ pub fn merge_commit_trees_without_repo(
                 .iter()
                 .map(|id| store.get_commit(id))
                 .try_collect()?;
-            let ancestor_tree = merge_commit_trees_without_repo(store, index, &ancestors)?;
+            let ancestor_tree =
+                merge_commit_trees_no_resolve_without_repo(store, index, &ancestors)?;
             let other_tree = other_commit.tree()?;
-            new_tree = new_tree.merge(&ancestor_tree, &other_tree)?;
+            new_tree = new_tree.merge_no_resolve(&ancestor_tree, &other_tree);
```
Comment (yuja) [inline lib/src/rewrite.rs:70]: @martinvonz Is this assumption valid?

> Since "ancestor_tree" isn't resolved by itself, this patch has subtle behavior
> change. For example, "jj diff -r9eaef582" in the "git" repository is no longer
> empty. I think the new behavior is also technically correct, but I'm not pretty
> sure.

Comment (martinvonz) [reply lib/src/rewrite.rs:70]: I think that may be due to the `A+(A-B)=A` rule resulting being applied in the ancestor before this patch (for those not familiar, that's https://github.com/martinvonz/jj/blob/bf2cb55ea7475adb36a93c8632979f956bbdedbf/lib/src/merge.rs#L92-L105 and also line 65 in that file). I've confirmed that removing that rule results in the same diff as with this patch (plus a ton of no-op "Resolved conflict" diffs). So I think in some sense the new behavior is more correct.

File: lib/src/merged_tree.rs
Comment (martinvonz) [inline lib/src/merged_tree.rs:1]: > This allows us to resolve conflicts of matching files only.

What do you mean by this?

FWIW, I've considered making the same change for #4152. Or maybe it was just in the context of that - that change will make it so we can more often simplify whole trees away.

Another advantage of it is that it reduces roundtrips to the backend by walking more trees in parallel. Without this change, we first walk from the roots to the differing leaves in the ancestor merge, then we walk again from the roots when merging the descendants (or other siblings in the case of octopus).

File: lib/src/rewrite.rs (hunk 1 above)
Comment (martinvonz) [reply lib/src/rewrite.rs:70]: Good find, btw!

File: lib/tests/test_revset.rs
Comment (martinvonz) [inline lib/tests/test_revset.rs:1]: nit: split this to a separate commit since it doesn't seem to depend on the changes in this commit

File: lib/src/default_index/revset_engine.rs
[the remark this answers is not in the document]
Comment (yuja) [reply lib/src/default_index/revset_engine.rs:1]: Updated, thanks.

File: lib/src/merged_tree.rs
Comment (yuja) [reply lib/src/merged_tree.rs:1]: Added code snippets. It's about `file()` revset.
```
merged_tree: allow to postpone resolution of intermediate trees

This allows us to diff trees without fully resolving conflicts:

    let from_tree = merge_no_resolve(..);
    for (path, (from, to)) in from_tree.diff(to_tree, matcher) {
        let from = resolve_conflicts(from);
        if from == to {
            continue; // resolved file may be identical
        ...
```

**Review by martinvonz — APPROVED** [review]
